numcosmo/galaxy/nc_galaxy_sd_obs_redshift_pz.c

```diff
@@ -0,0 +1,361 @@
+
+  self->sdz = NULL;
+}
+
+static void
+_nc_galaxy_sd_obs_redshift_pz_dispose (GObject *object)
+{
+  /* Chain up: end */
+  G_OBJECT_CLASS (nc_galaxy_sd_obs_redshift_pz_parent_class)->dispose (object);
+}
+
+static void
+nc_galaxy_sd_obs_redshift_pz_finalize (GObject *object)
+{
+  /* Chain up: end */
+  G_OBJECT_CLASS (nc_galaxy_sd_obs_redshift_pz_parent_class)->finalize (object);
+}
+
+static void _nc_galaxy_sd_obs_redshift_pz_gen (NcGalaxySDObsRedshift *gsdor, NcGalaxySDObsRedshiftData *data, NcmRNG *rng);
+static NcGalaxySDObsRedshiftIntegrand *_nc_galaxy_sd_obs_redshift_pz_integ (NcGalaxySDObsRedshift *gsdor);
+static void _nc_galaxy_sd_obs_redshift_pz_data_init (NcGalaxySDObsRedshift *gsdor, NcGalaxySDObsRedshiftData *data);
+static void _nc_galaxy_sd_obs_redshift_pz_add_submodel (NcmModel *model, NcmModel *submodel);
+
+static void
+nc_galaxy_sd_obs_redshift_pz_class_init (NcGalaxySDObsRedshiftPzClass *klass)
+{
+  NcGalaxySDObsRedshiftClass *gsdor_class = NC_GALAXY_SD_OBS_REDSHIFT_CLASS (klass);
+  GObjectClass *object_class              = G_OBJECT_CLASS (klass);
+  NcmModelClass *model_class              = NCM_MODEL_CLASS (klass);
+
+  object_class->dispose  = &_nc_galaxy_sd_obs_redshift_pz_dispose;
+  object_class->finalize = &nc_galaxy_sd_obs_redshift_pz_finalize;
+
+  ncm_model_class_set_name_nick (model_class, "P(z) Observed Redshift", "GalaxySDObsRedshiftPz");
+  ncm_model_class_add_params (model_class, 0, 0, PROP_LEN);
+  ncm_model_class_check_params_info (model_class);
+
+  gsdor_class->gen          = &_nc_galaxy_sd_obs_redshift_pz_gen;
+  gsdor_class->integ        = &_nc_galaxy_sd_obs_redshift_pz_integ;
+  gsdor_class->data_init    = &_nc_galaxy_sd_obs_redshift_pz_data_init;
+  model_class->add_submodel = &_nc_galaxy_sd_obs_redshift_pz_add_submodel;
+}
+
+static void
+_nc_galaxy_sd_obs_redshift_pz_gen (NcGalaxySDObsRedshift *gsdor, NcGalaxySDObsRedshiftData *data, NcmRNG *rng)
+{
+  NcGalaxySDObsRedshiftPz *gsdorpz            = NC_GALAXY_SD_OBS_REDSHIFT_PZ (gsdor);
+  NcGalaxySDObsRedshiftPzPrivate * const self = nc_galaxy_sd_obs_redshift_pz_get_instance_private (gsdorpz);
+  NcGalaxySDObsRedshiftPzData * const ldata   = (NcGalaxySDObsRedshiftPzData *) data->ldata;
+  guint npoints                               = 10000;
+  NcmVector *xv                               = ncm_vector_new (npoints);
+  NcmVector *yv                               = ncm_vector_new (npoints);
+  gdouble z_min                               = 0.0;
+  gdouble z_max                               = 0.0;
+  gdouble sigma                               = 0.05;
+  gdouble z;
+  guint i;
+
+  if (!nc_galaxy_sd_true_redshift_get_lim (self->sdz, &z_min, &z_max))
+    g_error ("Failed to get redshift limits.");
+
+  z     = nc_galaxy_sd_true_redshift_gen (self->sdz, rng);
+  sigma = (1.0 + z) * sigma;
+
+  for (i = 0; i < npoints; i++)
+  {
+    gdouble x = z_min + (z_max - z_min) * i / (npoints - 1);
```

Comment:
This should be generating a new true redshift from data->pz and leaving pz alone. In the future we can "resample" the distribution itself but it would be much more complicated than this.

Reply:
But this is used to generate new pz observations for numcosmo generate cluster-wl no? Without this, it isn't possible to generate them.

Reply:
It depends on what you are conditioning your statistics on. In this case, by doing what I'm saying we are keeping the photoz fixed and sampling only the true redshift. To sample the full photoz we need to condition the statistic to some distribution for pzs. The RAIL team in desc is doing that. But for the moment I think it is simpler and good approximation to keep the pz fixed.